File: cpp/cpp-std/demo12b01-data-race-single.cpp
/*
DATA RACES
Version 01: Without multithreading
*/


#include <iostream>
#include <vector>
#include <numeric>
using namespace std;



int getResult(int N) {
    vector<bool> a;
    a.resize(N + 1, false);

    for (int i = 1; i <= N; ++i)
        if (0 == i % 2 || 0 == i % 3)
            a[i] = true;

    // result = sum of a (i.e. counting number of true values in a)
    int result = std::accumulate(a.begin(), a.end(), 0);
    return result;
}



int main() {
    constexpr int N = 8;

    int result = getResult(N);

    cout << "Number of integers that are divisible by 2 or 3 is: " << result << endl;
    return 0;
}

File: js-nodejs/demo12b01-data-race-single.js
/*
DATA RACES
Version 01: Without multithreading
*/

const getResult = N => {
  const a = new Int8Array(N + 1).fill(0);
  for (let i = 1; i <= N; ++i) {
    if (i % 2 === 0 || i % 3 === 0)
      a[i] = 1;
  }
  const result = a.reduce((x, y) => x + y, 0);
  return result;
};


const mainFunc = () => {
  const N = 1024;
  const result = getResult(N);
  console.log('Number of integers that are divisible by 2 or 3 is:', result);
};


mainFunc();
